src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelBindingContext.cs

```diff
@@ -232,69 +82,93 @@ public ModelStateDictionary ModelState
         /// </summary>
-        /// <remarks>Passed into the model binding system.</remarks>
-        public bool IsTopLevelObject { get; set; }
+        public abstract ModelBindingResult? Result { get; set; }
 
         /// <summary>
-        /// Gets or sets the <see cref="IValueProvider"/> associated with this context.
+        /// Pushes a layer of state onto this context. Model binders will call this as part of recursion when binding properties
+        /// or collection items.
         /// </summary>
-        public IValueProvider ValueProvider
-        {
-            get { return _valueProvider; }
-            set
-            {
-                if (value == null)
-                {
-                    throw new ArgumentNullException(nameof(value));
-                }
+        /// <param name="modelMetadata"><see cref="ModelBinding.ModelMetadata"/> to assign to the <see cref="ModelMetadata"/> property</param>
+        /// <param name="fieldName">Name to assign to the <see cref="FieldName"/> property</param>
+        /// <param name="modelName">Name to assign to the <see cref="ModelName"/> property</param>
+        /// <param name="model">Instance to assign to the <see cref="Model"/> property</param>
+        /// <returns>A <see cref="NestedScope"/> scope object which should be used in a using statement where PushContext is called.</returns>
+        public abstract NestedScope EnterNestedScope(ModelMetadata modelMetadata, string fieldName, string modelName, object model);
 
-                _valueProvider = value;
-            }
-        }
+        /// <summary>
+        /// Pushes a layer of state onto this context. Model binders will call this as part of recursion when binding properties
+        /// or collection items.
+        /// </summary>
+        /// <returns>A <see cref="NestedScope"/> scope object which should be used in a using statement where PushContext is called.</returns>        
+        public abstract NestedScope EnterNestedScope();
 
         /// <summary>
-        /// Gets or sets a predicate which will be evaluated for each property to determine if the property
-        /// is eligible for model binding.
+        /// Removes a layer of state pushed by calling <see cref="EnterNestedScope"/>.
         /// </summary>
-        public Func<ModelBindingContext, string, bool> PropertyFilter { get; set; }
+        protected abstract void ExitNestedScope();
 
         /// <summary>
-        /// Gets or sets the <see cref="ValidationStateDictionary"/>. Used for tracking validation state to
-        /// customize validation behavior for a model object.
+        /// Return value of <see cref="ModelBindingContext.EnterNestedScope"/>. Should be disposed
+        /// by caller when child binding context state should be popped off of 
+        /// the <see cref="ModelBindingContext"/>.
         /// </summary>
-        public ValidationStateDictionary ValidationState { get; set; }
+        public struct NestedScope : IDisposable
```

Comment:
@rynowak why did you suggest nesting this class? It has to be `public` to avoid exposure as just another `IDisposable`.

Or does the compiler check the runtime type when it gens for a `using`, allowing this to be `private`?